snapcraft/tests/plugins/python/test_python_finder.py

```diff
@@ -0,0 +1,114 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import os
+import re
+
+from testtools.matchers import (
+    Equals,
+    MatchesRegex,
+)
+
+from snapcraft.plugins._python import _python_finder
+
+from snapcraft import (
+    tests
+)
+
+
+class GetPythonCommandTestCase(tests.TestCase):
+
+    def test_staged(self):
+        """get_python_command should support staged python"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create staged binary
+        _create_python_binary(stage_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                stage_dir, 'usr', 'bin', 'python2')))
+
+    def test_installed(self):
+        """get_python_command should support installed python"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create installed binary
+        _create_python_binary(install_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                install_dir, 'usr', 'bin', 'python2')))
+
+    def test_staged_and_installed(self):
+        """get_python_command should prefer staged python over installed"""
+
+        stage_dir = 'stage_dir'
+        install_dir = 'install_dir'
+
+        # Create both staged and installed binaries
+        _create_python_binary(stage_dir)
+        _create_python_binary(install_dir)
+
+        python_command = _python_finder.get_python_command(
+            'python2', stage_dir, install_dir)
+        self.assertThat(
+            python_command, Equals(os.path.join(
+                stage_dir, 'usr', 'bin', 'python2')))
+
+
+class GetPythonHeadersTestCase(tests.TestCase):
+
+    def test_staged(self):
+        """get_python_headers should support staged headers"""
+
+        stage_dir = 'stage_dir'
+
+        # Create staged headers
+        _create_python_headers(stage_dir)
+
+        self.assertThat(
+            _python_finder.get_python_headers('python2', stage_dir),
+            Equals(os.path.join(stage_dir, 'usr', 'include', 'python2')))
+
+    def test_unstaged(self):
+        """get_python_headers should support unstaged headers"""
+
+        stage_dir = 'stage_dir'
+
+        self.assertThat(
+            _python_finder.get_python_headers('python3', stage_dir),
+            MatchesRegex(re.escape(os.path.join(
+                os.sep, 'usr', 'include', 'python3'))))
+
+
+def _create_python_binary(base_dir):
+    python_command_path = os.path.join(
+        base_dir, 'usr', 'bin', 'python2')
+    os.makedirs(os.path.dirname(python_command_path))
+    open(python_command_path, 'w').close()
+
+
+def _create_python_headers(base_dir):
+    os.makedirs(os.path.join(base_dir, 'usr', 'include', 'python2'))
```

Comment:
I prefer to put the helpers at the start of the file the things that are global. I thought that was pep8, but can't find a specific mention to this. Why do you like them at the bottom?

Reply:
You're right, I'm sorry, I keep forgetting about this. I like them at the bottom because it seems inherently natural that the public functions are more interesting to the reader, thus should be closer to the top. It's like designing a web-page: put the stuff people want to see front and center. But I'm happy to switch the order.

Reply:
Done.

Reply:
That's what had me confused. I like that rule of more public first, and then group the helpers after they are used. Maybe, they removed the pep8 part that talked about organizing the functions in the module, and now I have an internal conflict :)

Reply:
Ha! Well, think on it. Super easy to flip around.